design-proposals/shadow-node.md

```diff
@@ -0,0 +1,210 @@
+Virt handler can use its existing [web-service](https://github.com/kubevirt/kubevirt/blob/b102c56f0fcd52feff3ff7a6296737b8e8b99131/cmd/virt-handler/virt-handler.go#L556) to expose one more endpoint - GetNodeInfo. Once reached - this endpoint will return the node data.
+
+Virt-Controller (Client): Add a new web-service client infra that will:
+1. run as a go routine and poll the client for updates every one minute.This frequency is chosen because it is the highest of the frequencies (=heartbeat).
+2. ask every Virt-handler server separately for the NodeInfo (need to make sure it's possible to address them in a unique manner).
+3. filter non-kubevirt owned annotations/labels.
+4. patch the node if necessary.
+
+The heartbeat annotation and virt-controller's [Node-controller](https://github.com/kubevirt/kubevirt/blob/b102c56f0fcd52feff3ff7a6296737b8e8b99131/pkg/virt-controller/watch/node.go) can be removed in favor for the web-service timeout mechanism (if received timeout - set node as unreachable).
+
+Pros:
+- Avoid introducing a new CRD to the project.
+- Using existing web-server infra on the server side.
+- security:
+  - the web-service is secure using the cert-manager internal API.
+  - Virt-handler does not need to access kube-API at all, and cannot modify unrelated nodes or shadow nodes. Virt-controller does all the risky work
+- Scalability:
+  - Reduces #API requests (i.e. node patches) to a minimum.
+
+Cons:
+- More complex. It's easier and more appealing to use the k8s API, i.e. patching labels/annotations, than to implement and maintain GRPC/REST case logic internally.
+- Virt-handler will no longer be able to immediately set the node as unschedulable (in case of virt-handler SIGTERM). However, this use-case is pretty rare, and IMO should be reconsidered.
+
+## Solution option #3: Moving to third party entity
+The idea is to eliminate the reason(s) to patch to the node from virt-handler in the first place.
+
+1. Use tools like [Node Feature Discovery](https://kubernetes-sigs.github.io/node-feature-discovery/v0.15/get-started/introduction.html) or utilize virt-handler's [device-manager](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/) to publish node features (pending checking whether they don't have the CVE issue as well).
+2. Remove writing the unschedulable label in case virt-handler crashes. It is racy, and has no real value.
+3. Heartbeat - Remove adding the annotation in favor of watching the readiness of the virt-handler Deamonset/pods.
+
+Pros:
+- No need to invent the wheel.
+- Re-examine current methods to make sure if we still need them is a good practice. In this case we may find out that patching is no longer needed, and in the long run need to maintain less code.
+- This steps can be broken to multiple seemingly unrelated PRs, that are easier to review.
+
+Cons:
+- Relying on a third part tools.
```

Comment:
I don't understand this statement: afaict KubeVirt already depends on Node Feature Discovery and Device Plugins.

Reply:
virt-handler uses device-manger, but AFIK not NFD:
```
$ git grep node-feature-discovery
tests/vmi_lifecycle_test.go:                            By("adding a node-feature-discovery CPU model label to a node")
```

Reply:
kubevirt is compatible with NFD, but all virt specific discovery, PCI device discovery and management, and much more is all done in kubevirt.

I don't think NFD is an option, and by moving it from an (initially external approach) into kubevirt we simplified the architecture, increase feature velocity and removed a lot of external failure points. Would advise against externalizing anything in that area, while of course not doing anything which would make kubevirt not being able to pick up NFD stuff in addition.

Reply:
ACK. + NFD will not filter kubevirt-owned labels.

Reply:
I agree with Roman, labels are too critical for KubeVirt in order to rely on an external project such as NFD.

Reply:
+1 in regards that moving this to NFD would be step back and on top of it this option seems like a lot of work (at least in comparison to other options)

Reply:
ACK